cassandra/pool.py

```diff
@@ -450,36 +476,48 @@ def shutdown(self):
                 self.is_shutdown = True
             self._stream_available_condition.notify_all()
 
-        if self._connection:
-            self._connection.close()
-            self._connection = None
+        if self._connections:
+            for c in self._connections.values():
+                c.close()
+            self._connections = dict()
 
     def _set_keyspace_for_all_conns(self, keyspace, callback):
-        if self.is_shutdown or not self._connection:
+        """
+        Asynchronously sets the keyspace for all connections.  When all
+        connections have been set, `callback` will be called with two
+        arguments: this pool, and a list of any errors that occurred.
+        """
+        remaining_callbacks = set(self._connections.values())
+        errors = []
+
+        if not remaining_callbacks:
+            callback(self, errors)
             return
 
         def connection_finished_setting_keyspace(conn, error):
             self.return_connection(conn)
-            errors = [] if not error else [error]
-            callback(self, errors)
+            remaining_callbacks.remove(conn)
```

Comment:
Are you sure that `set` is a thread safe object? 
Note that set.remove() is called from the async() context.

Reply:
Set append/remove are thread safe (cause of GIL), and in our case it would be call exactly once for each connection, hence there won't be multiple calls to remove the same connection.

I've copied the logic from `HostConnectionPool._set_keyspace_for_all_conns` which have the same semantic of handling multiple connections.

So my assumption it should solid enough.

Reply:
I'm not sure you can rely on this.
Read some discussion about this here: https://stackoverflow.com/questions/2227169/are-python-built-in-containers-thread-safe

Reply:
ping. I think we'd rather play it safe here.

Reply:
we could switch to a queue here if we want to be 100% safe; or use a lock

Reply:
we ended up with a Lock